src/crimson/thread/ThreadPool.cc

```diff
@@ -0,0 +1,77 @@
+#include <chrono>
+
+#include "ThreadPool.h"
+#include "crimson/net/Config.h"
+
+namespace ceph::thread {
+
+ThreadPool::ThreadPool(size_t n)
+  : threads{create_threads(n)}
+{
+  // try to shard the queue evenly
+  const auto queue_size = num_queue_size / seastar::smp::count;
+  for (size_t i = 0; i < seastar::smp::count; i++) {
+    if (i == 0) {
+      free_slots.emplace_back(queue_size + num_queue_size % queue_size);
+    } else {
+      free_slots.emplace_back(queue_size);
+    }
+  }
+}
+
+ThreadPool::~ThreadPool()
+{
+  stop = true;
+  cond.notify_all();
+  for (auto& thread : threads) {
+    thread.join();
```

Comment:
do we need a way to do async shutdown? a blocking destructor means we can't create and destroy these at runtime in seastar threads, so any of these pools have to live on the stack in `main()`. that's a pretty big limitation, but may be okay?

Reply:
why would we need to create and destroy these in seastar threads? yeah, it's limitation. but i want to know the use case of managing the life cycle of alien thread pools in seastar.

i don't really think they will live on *stack*, they would be created on heap and managed by a smart ptr.

Reply:
as an example, if we're planning to host multiple logical OSDs in a single process, we may be dynamically starting them up/shutting them down at runtime. if each has a separate thread pool, we'd be managing them from within seastar

but given that a) we'll be carefully pinning these threads to non-seastar cores, and b) ThreadPool::submit() is completely generic, i expect that one or more global pools will be good enough. so putting them on the stack in main() seems reasonable to me - but using a shared ptr sounds dangerous, as you might drop the last reference inside a seastar thread

Reply:
> but using a shared ptr sounds dangerous, as you might drop the last reference inside a seastar thread

oic. i misunderstood your question. i was thinking about a `unique_ptr<>` in `main()`.

Reply:
i think we should just let these OSD instances share the same thread pool, i think that's the merit of hosting multiple OSD instance in the same process. unless we want to throttle the usage of alien threads by OSD. but i think we can always achieve the same goal at a different phase when serving a request.